deep_speech_2/train.py

```diff
@@ -0,0 +1,128 @@
+    audio_data = paddle.layer.data(
+        name="audio_spectrogram",
+        height=161,
+        width=1000,
+        type=paddle.data_type.dense_vector(161000))
+    text_data = paddle.layer.data(
+        name="transcript_text",
+        type=paddle.data_type.integer_value_sequence(dict_size))
+    cost, _ = deep_speech2(
+        audio_data=audio_data,
+        text_data=text_data,
+        dict_size=dict_size,
+        num_conv_layers=args.num_conv_layers,
+        num_rnn_layers=args.num_rnn_layers,
+        rnn_size=args.rnn_layer_size)
+
+    # create parameters and optimizer
+    parameters = paddle.parameters.create(cost)
+    optimizer = paddle.optimizer.Adam(
+        learning_rate=5e-4, gradient_clipping_threshold=400)
+    trainer = paddle.trainer.SGD(
+        cost=cost, parameters=parameters, update_equation=optimizer)
+    # create data readers
+    feeding = {
+        "audio_spectrogram": 0,
+        "transcript_text": 1,
+    }
+    train_batch_reader_with_sortagrad = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.train", sort_by_duration=True),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+    train_batch_reader_without_sortagrad = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.train",
+                sort_by_duration=False,
+                shuffle=True),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+    test_batch_reader = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.dev", sort_by_duration=False),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+
+    # create event handler
+    def event_handler(event):
+        if isinstance(event, paddle.event.EndIteration):
+            if event.batch_id % 10 == 0:
+                print "/nPass: %d, Batch: %d, TrainCost: %f" % (
+                    event.pass_id, event.batch_id, event.cost)
+            else:
+                sys.stdout.write('.')
+                sys.stdout.flush()
+        if isinstance(event, paddle.event.EndPass):
+            result = trainer.test(reader=test_batch_reader, feeding=feeding)
+            print "Pass: %d, TestCost: %s" % (event.pass_id, result.cost)
+            with gzip.open("params.tar.gz", 'w') as f:
+                parameters.to_tar(f)
+
+    # run train
+    # first pass with sortagrad
+    if args.use_sortagrad:
+        trainer.train(
+            reader=train_batch_reader_with_sortagrad,
+            event_handler=event_handler,
+            num_passes=1,
+            feeding=feeding)
+        args.num_passes -= 1
+    # other passes without sortagrad
+    trainer.train(
```

Comment:
I think more elaborate controlling interfaces of training process need to be exposed to users for more flexible training flow control. E.g. in this DS2 case, training data needs to be changed during training. In other cases, parts of model needs freezing for a while or being trained alternatively (e.g. GAN).